### PR TITLE
[fix](memory) Fix Allocator release memory to correct MemTracker after TLS attach task ends

### PR DESCRIPTION
**Pull request by xinyiZzz**
Allocator save TLS MemTracker during first alloc, which is used to release memory after TLS attach task ends.
```
23:00:15   F20240824 22:56:49.773799 66432 thread_context.h:238] Check failed: doris::k_doris_exit || !doris::config::enable_memory_orphan_check || thread_mem_tracker()->label() != "Orphan" If you crash here, it means that SCOPED_ATTACH_TASK and SCOPED_SWITCH_THREAD_MEM_TRACKER_LIMITER are not used correctly. starting position of each thread is expected to use SCOPED_ATTACH_TASK to bind a MemTrackerLimiter belonging to Query/Load/Compaction/Other Tasks, otherwise memory alloc using Doris Allocator in the thread will crash. If you want to switch MemTrackerLimiter during thread execution, please use SCOPED_SWITCH_THREAD_MEM_TRACKER_LIMITER, do not repeat Attach. Of course, you can modify enable_memory_orphan_check=false in be.conf to avoid this crash.
23:00:15   *** Check failure stack trace: ***
23:00:15       @     0x55645d3388e6  google::LogMessageFatal::~LogMessageFatal()
23:00:15       @     0x5564439637c2  doris::ThreadContext::consume_memory()
23:00:15       @     0x5564439914fe  Allocator<>::release_memory()
23:00:15       @     0x5564354be11e  std::_Sp_counted_ptr<>::_M_dispose()
23:00:15       @     0x55643557bc3b  std::deque<>::pop_front()
23:00:15       @     0x5564355756b1  doris::io::StreamLoadPipe::~StreamLoadPipe()
23:00:15       @     0x5564354bfa77  doris::StreamLoadContext::~StreamLoadContext()
23:00:15       @     0x556436ee5114  doris::HttpRequest::~HttpRequest()
```

### CODE REVIEW

**Comment by xinyiZzz** [comment]
run buildall

File: be/src/common/config.cpp [hunk 1]
```diff
@@ -142,7 +142,7 @@ DEFINE_mInt64(stacktrace_in_alloc_large_memory_bytes, "2147483648");
 
 DEFINE_mInt64(crash_in_alloc_large_memory_bytes, "-1");
 
-DEFINE_mBool(enable_memory_orphan_check, "false");
+DEFINE_mBool(enable_memory_orphan_check, "true");
```
Comment (yiguolei) [inline be/src/common/config.cpp:145]: add comment here not in .h file,

File: be/src/common/config.h [hunk 2]
```diff
@@ -197,6 +197,7 @@ DECLARE_mInt64(stacktrace_in_alloc_large_memory_bytes);
 DECLARE_mInt64(crash_in_alloc_large_memory_bytes);
 
 // default is true. if any memory tracking in Orphan mem tracker will report error.
+// not modify the default value of this conf!! otherwise memory errors cannot be detected in time.
```
Comment (yiguolei) [inline be/src/common/config.h:200]: It is useless to add comment here.

Comment (xinyiZzz) [reply be/src/common/config.h]: both add

**Comment by xinyiZzz** [comment]
run buildall

**Comment by xinyiZzz** [comment]
run buildall

**Review by yiguolei — DISMISSED** [review]

File: be/src/vec/common/allocator.h [hunk 3]
```diff
@@ -395,6 +399,8 @@ class Allocator {
         return buf;
     }
 
+    std::shared_ptr<doris::MemTrackerLimiter> mem_tracker {nullptr};
```
[the remark this answers is not in the document]
Comment (xinyiZzz) [reply be/src/vec/common/allocator.h:402]: done

File: be/src/vec/common/allocator.cpp [hunk 4]
```diff
@@ -211,14 +211,44 @@ void Allocator<clear_memory_, mmap_populate, use_mmap, MemoryAllocator>::memory_
...
+        // Allocator tracker label, it usually happens during object destruction. This means that
+        // the scope of SCOPED_ATTACH_TASK has been left,  so release memory using Allocator tracker.
+        SCOPED_SWITCH_THREAD_MEM_TRACKER_LIMITER(mem_tracker);
+        RELEASE_THREAD_MEM_TRACKER(size);
+    } else {
+        // there is a scenario where an object that inherits Allocator has never called alloc, but frees memory.
+        // in phmap, the memory alloced by an object may be transferred to another object and then free.
+        // in this case, thread context must attach a memory tracker other than Orphan,
+        // otherwise memory tracking will be wrong.
+        DCHECK(thread_context && thread_context->thread_mem_tracker()->label() != "Orphan");
```
[the remark this answers is not in the document]
Comment (xinyiZzz) [reply be/src/vec/common/allocator.cpp:249]: 有可能，所以会走到第二条路径，使用 allocator 里保存的 tracker
image
第三条路径的意思是，如果 allocator tracker == nullptr，那 DCHECK thread context  一定不等于 nullptr

**Comment by xinyiZzz** [comment]
TODO dcheck load
DCHECK(doris::thread_context()->thread_mem_tracker()->type() != doris::MemTrackerLimiter::Type::QUERY || doris::thread_context()->thread_mem_tracker()->label() == mem_tracker_->label())
                << ", thread mem tracker label: "
                << doris::thread_context()->thread_mem_tracker()->label()
                << ", allocator tracker label: " << mem_tracker_->label();
```
F20240826 21:01:08.522873 135505 allocator.cpp:220] Check failed: doris::thread_context()->thread_mem_tracker()->label() == mem_tracker->label() , thread mem tracker label: Load#Id=6834273cfabe49cb-9d783fde25a0d61a, allocator tracker label: 8#0
21:11:13   *** Check failure stack trace: ***
21:11:13       @     0x55a5a4d2a0d6  google::LogMessageFatal::~LogMessageFatal()
21:11:13       @     0x55a5891a3a7d  Allocator<>::consume_memory()
21:11:13       @     0x55a5891a6213  Allocator<>::alloc_impl()
21:11:13       @     0x55a5a72df17f  orc::DataBuffer<>::reserve()
21:11:13       @     0x55a5a72de421  orc::DataBuffer<>::resize()
21:11:13       @     0x55a5a72def7f  orc::DataBuffer<>::DataBuffer()
21:11:13       @     0x55a5a72f3723  std::make_unique<>()
21:11:13       @     0x55a5a72edfcf  orc::createReader()
21:11:13       @     0x55a58969edcb  doris::vectorized::OrcReader::_create_file_reader()
21:11:13       @     0x55a5896a0947  doris::vectorized::OrcReader::init_reader()
21:11:13       @     0x55a5a300b14c  doris::vectorized::VFileScanner::_get_next_reader()
21:11:13       @     0x55a5a3001e92  doris::vectorized::VFileScanner::_get_block_wrapped()
21:11:13       @     0x55a5a3000f5a  doris::vectorized::VFileScanner::_get_block_impl()
21:11:13       @     0x55a58a296eb9  doris::vectorized::VScanner::get_block()
21:11:13       @     0x55a58a2962f5  doris::vectorized::VScanner::get_block_after_projects()
21:11:13       @     0x55a58a248798  doris::vectorized::ScannerScheduler::_scanner_scan()
21:11:13       @     0x55a58a24c1e0  _ZNSt17_Function_handlerIFvvEZZN5doris10vectorized16ScannerScheduler6submitESt10shared_ptrINS2_14ScannerContextEES4_INS2_8ScanTaskEEENK3$_1clEvEUlvE_E9_M_invokeERKSt9_Any_data
21:11:13       @     0x55a57316801e  doris::ThreadPool::dispatch_thread()
21:11:13       @     0x55a57314125f  doris::Thread::supervise_thread()
21:11:13       @     0x7f8dc7b17609  start_thread
21:11:13       @     0x7f8dc7dc4133  clone
21:11:13       @              (nil)  (unknown)
21:11:13   *** Query id: 6834273cfabe49cb-9d783fde25a0d61a ***
21:11:13   *** is nereids: 1 ***
21:11:13   *** tablet id: 0 ***
21:11:13   *** Aborted at 1724677269 (unix time) try "date -d @1724677269" if you are using GNU date ***
21:11:13   *** Current BE git commitID: fbd4a0087c ***
21:11:13   *** SIGABRT unknown detail explain (@0x209f7) received by PID 133623 (TID 135505 OR 0x7f8627a04700) from PID 133623; stack trace: ***
21:11:13    0# doris::signal::(anonymous namespace)::FailureSignalHandler(int, siginfo_t*, void*) at /root/doris/be/src/common/signal_handler.h:421
21:11:13    1# 0x00007F8DC7CE8090 in /lib/x86_64-linux-gnu/libc.so.6
21:11:13    2# raise at ../sysdeps/unix/sysv/linux/raise.c:51
21:11:13    3# abort at /build/glibc-SzIz7B/glibc-2.31/stdlib/abort.c:81
21:11:13    4# 0x000055A5A4D3104D in /mnt/ssd01/pipline/OpenSourceDoris/clusterEnv/P0/Cluster0/be/lib/doris_be
21:11:13    5# google::LogMessage::SendToLog() in /mnt/ssd01/pipline/OpenSourceDoris/clusterEnv/P0/Cluster0/be/lib/doris_be
21:11:13    6# google::LogMessage::Flush() in /mnt/ssd01/pipline/OpenSourceDoris/clusterEnv/P0/Cluster0/be/lib/doris_be
21:11:13    7# google::LogMessageFatal::~LogMessageFatal() in /mnt/ssd01/pipline/OpenSourceDoris/clusterEnv/P0/Cluster0/be/lib/doris_be
21:11:13    8# Allocator<false, false, false, RecordSizeMemoryAllocator>::consume_memory(unsigned long) at /root/doris/be/src/vec/common/allocator.cpp:220
21:11:13    9# Allocator<false, false, false, RecordSizeMemoryAllocator>::alloc_impl(unsigned long, unsigned long) at /root/doris/be/src/vec/common/allocator.h:251
21:11:13   10# orc::DataBuffer<char>::reserve(unsigned long) in /mnt/ssd01/pipline/OpenSourceDoris/clusterEnv/P0/Cluster0/be/lib/doris_be
21:11:13   11# orc::DataBuffer<char>::resize(unsigned long) in /mnt/ssd01/pipline/OpenSourceDoris/clusterEnv/P0/Cluster0/be/lib/doris_be
21:11:13   12# orc::DataBuffer<char>::DataBuffer(orc::MemoryPool&, unsigned long) in /mnt/ssd01/pipline/OpenSourceDoris/clusterEnv/P0/Cluster0/be/lib/doris_be
21:11:13   13# std::_MakeUniq<orc::DataBuffer<char> >::__single_object std::make_unique<orc::DataBuffer<char>, orc::MemoryPool&, unsigned long&>(orc::MemoryPool&, unsigned long&) in /mnt/ssd01/pipline/OpenSourceDoris/clusterEnv/P0/Cluster0/be/lib/doris_be
21:11:13   14# orc::createReader(std::unique_ptr<orc::InputStream, std::default_delete<orc::InputStream> >, orc::ReaderOptions const&) in /mnt/ssd01/pipline/OpenSourceDoris/clusterEnv/P0/Cluster0/be/lib/doris_be
21:11:13   15# doris::vectorized::OrcReader::_create_file_reader() in /mnt/ssd01/pipline/OpenSourceDoris/clusterEnv/P0/Cluster0/be/lib/doris_be
21:11:13   16# doris::vectorized::OrcReader::init_reader(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const*, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::variant<doris::ColumnValueRange<(doris::PrimitiveType)3>, doris::ColumnValueRange<(doris::PrimitiveType)4>, doris::ColumnValueRange<(doris::PrimitiveType)5>, doris::ColumnValueRange<(doris::PrimitiveType)6>, doris::ColumnValueRange<(doris::PrimitiveType)7>, doris::ColumnValueRange<(doris::PrimitiveType)36>, doris::ColumnValueRange<(doris::PrimitiveType)37>, doris::ColumnValueRange<(doris::PrimitiveType)15>, doris::ColumnValueRange<(doris::PrimitiveType)10>, doris::ColumnValueRange<(doris::PrimitiveType)23>, doris::ColumnValueRange<(doris::PrimitiveType)11>, doris::ColumnValueRange<(doris::PrimitiveType)25>, doris::ColumnValueRange<(doris::PrimitiveType)12>, doris::ColumnValueRange<(doris::PrimitiveType)26>, doris::ColumnValueRange<(doris::PrimitiveType)20>, doris::ColumnValueRange<(doris::PrimitiveType)2>, doris::ColumnValueRange<(doris::PrimitiveType)19>, doris::ColumnValueRange<(doris::PrimitiveType)28>, doris::ColumnValueRange<(doris::PrimitiveType)29>, doris::ColumnValueRange<(doris::PrimitiveType)30>, doris::ColumnValueRange<(doris::PrimitiveType)35> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::variant<doris::ColumnValueRange<(doris::PrimitiveType)3>, doris::ColumnValueRange<(doris::PrimitiveType)4>, doris::ColumnValueRange<(doris::PrimitiveType)5>, doris::ColumnValueRange<(doris::PrimitiveType)6>, doris::ColumnValueRange<(doris::PrimitiveType)7>, doris::ColumnValueRange<(doris::PrimitiveType)36>, doris::ColumnValueRange<(doris::PrimitiveType)37>, doris::ColumnValueRange<(doris::PrimitiveType)15>, doris::ColumnValueRange<(doris::PrimitiveType)10>, doris::ColumnValueRange<(doris::PrimitiveType)23>, doris::ColumnValueRange<(doris::PrimitiveType)11>, doris::ColumnValueRange<(doris::PrimitiveType)25>, doris::ColumnValueRange<(doris::PrimitiveType)12>, doris::ColumnValueRange<(doris::PrimitiveType)26>, doris::ColumnValueRange<(doris::PrimitiveType)20>, doris::ColumnValueRange<(doris::PrimitiveType)2>, doris::ColumnValueRange<(doris::PrimitiveType)19>, doris::ColumnValueRange<(doris::PrimitiveType)28>, doris::ColumnValueRange<(doris::PrimitiveType)29>, doris::ColumnValueRange<(doris::PrimitiveType)30>, doris::ColumnValueRange<(doris::PrimitiveType)35> > > > >*, std::vector<std::shared_ptr<doris::vectorized::VExprContext>, std::allocator<std::shared_ptr<doris::vectorized::VExprContext> > > const&, bool, doris::TupleDescriptor const*, doris::RowDescriptor const*, std::vector<std::shared_ptr<doris::vectorized::VExprContext>, std::allocator<std::shared_ptr<doris::vectorized::VExprContext> > > const*, std::unordered_map<int, std::vector<std::shared_ptr<doris::vectorized::VExprContext>, std::allocator<std::shared_ptr<doris::vectorized::VExprContext> > >, std::hash<int>, std::equal_to<int>, std::allocator<std::pair<int const, std::vector<std::shared_ptr<doris::vectorized::VExprContext>, std::allocator<std::shared_ptr<doris::vectorized::VExprContext> > > > > > const*, bool) at /root/doris/be/src/vec/exec/format/orc/vorc_reader.cpp:300
21:11:13   17# doris::vectorized::VFileScanner::_get_next_reader() at /root/doris/be/src/vec/exec/scan/vfile_scanner.cpp:906
21:11:13   18# doris::vectorized::VFileScanner::_get_block_wrapped(doris::RuntimeState*, doris::vectorized::Block*, bool*) in /mnt/ssd01/pipline/OpenSourceDoris/clusterEnv/P0/Cluster0/be/lib/doris_be
21:11:13   19# doris::vectorized::VFileScanner::_get_block_impl(doris::RuntimeState*, doris::vectorized::Block*, bool*) at /root/doris/be/src/vec/exec/scan/vfile_scanner.cpp:253
21:11:13   20# doris::vectorized::VScanner::get_block(doris::RuntimeState*, doris::vectorized::Block*, bool*) in /mnt/ssd01/pipline/OpenSourceDoris/clusterEnv/P0/Cluster0/be/lib/doris_be
21:11:13   21# doris::vectorized::VScanner::get_block_after_projects(doris::RuntimeState*, doris::vectorized::Block*, bool*) at /root/doris/be/src/vec/exec/scan/vscanner.cpp:81
21:11:13   22# doris::vectorized::ScannerScheduler::_scanner_scan(std::shared_ptr<doris::vectorized::ScannerContext>, std::shared_ptr<doris::vectorized::ScanTask>) at /root/doris/be/src/vec/exec/scan/scanner_scheduler.cpp:236
21:11:13   23# std::_Function_handler<void (), doris::vectorized::ScannerScheduler::submit(std::shared_ptr<doris::vectorized::ScannerContext>, std::shared_ptr<doris::vectorized::ScanTask>)::$_1::operator()() const::{lambda()#1}>::_M_invoke(std::_Any_data const&) at /var/local/ldb-toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/std_function.h:291
21:11:13   24# doris::ThreadPool::dispatch_thread() at /root/doris/be/src/util/threadpool.cpp:551
21:11:13   25# doris::Thread::supervise_thread(void*) at /root/doris/be/src/util/thread.cpp:499
21:11:13   26# start_thread at /build/glibc-SzIz7B/glibc-2.31/nptl/pthread_create.c:478
21:11:13   27# __clone at ../sysdeps/unix/sysv/linux/x86_64/clone.S:97
```

**Comment by xinyiZzz** [comment]
run buildall

**Comment by xinyiZzz** [comment]
run buildall

**Comment by xinyiZzz** [comment]
run buildall

**Comment by xinyiZzz** [comment]
run buildall

**Review by yiguolei — APPROVED** [review]

**Comment by xinyiZzz** [comment]
run buildall

**Review by yiguolei — APPROVED** [review]

**Review by jacktengg — APPROVED** [review]

**Review by yiguolei — APPROVED** [review]

**Review by gavinchou — APPROVED** [review]

**Pull request merged**